stable_baselines3/common/buffers.py

```diff
@@ -387,15 +380,81 @@ def get(self, batch_size: Optional[int] = None) -> Generator[RolloutBufferSample
-                     env: Optional[VecNormalize] = None) -> RolloutBufferSamples:
-        data = (self.observations[batch_inds],
-                self.actions[batch_inds],
-                self.values[batch_inds].flatten(),
-                self.log_probs[batch_inds].flatten(),
-                self.advantages[batch_inds].flatten(),
-                self.returns[batch_inds].flatten())
+    def _get_samples(self, batch_inds: np.ndarray, env: Optional[VecNormalize] = None) -> RolloutBufferSamples:
+        data = (
+            self.observations[batch_inds],
+            self.actions[batch_inds],
+            self.values[batch_inds].flatten(),
+            self.log_probs[batch_inds].flatten(),
+            self.advantages[batch_inds].flatten(),
+            self.returns[batch_inds].flatten(),
+        )
         return RolloutBufferSamples(*tuple(map(self.to_torch, data)))
+
+
+class NstepReplayBuffer(ReplayBuffer):
+    def __init__(
+        self,
+        buffer_size: int,
+        observation_space: spaces.Space,
+        action_space: spaces.Space,
+        device: Union[th.device, str] = "cpu",
+        n_envs: int = 1,
+        optimize_memory_usage: bool = False,
+        n_step: int = 1,
+        gamma: float = 0.99,
+    ):
+        super().__init__(buffer_size, observation_space, action_space, device, n_envs, optimize_memory_usage)
+        self.n_step = int(n_step)
+        assert self.n_step <= buffer_size
+        self.gamma = float(gamma)
+        assert 0 <= gamma <= 1
+
+    def _get_samples(self, batch_inds: np.ndarray, env: Optional[VecNormalize] = None) -> ReplayBufferSamples:
+        assert not np.any(batch_inds == self.pos)
+        actions = self.actions[batch_inds, 0, :]
+
+        gamma = self.gamma
+
+        # Broadcasting turns 1dim arange matrix to 2 dimensional matrix that contains all
+        # the indices, % buffersize keeps us in buffer range
+        indices = (np.arange(self.n_step) + batch_inds.reshape(-1, 1)) % self.buffer_size
+
+        # two dim matrix of not dones. If done is true, then subsequent dones are turned to 0
+        # using accumulate. This ensures that we don't use invalid transitions
+        not_dones = np.multiply.accumulate(1 - self.dones[indices], 1).reshape(-1, self.n_step)
+
+        # vector of the discount factors
+        gammas = gamma ** np.arange(self.n_step)
+
+        # two dim matrix of rewards for the indices
+        rewards = self.rewards[indices].reshape(not_dones.shape)
+        rewards = self._normalize_reward(rewards, env)
+
+        # filter to ignore loops around the buffer, see:
+        # https://github.com/DLR-RM/stable-baselines3/pull/81#issuecomment-653741592
+        # basically we need to ignore indices that are equal to self.pos
+        # because the indice at self.pos is older
+        # not_dones that are from other transitions are filtered out.
+        filt = np.hstack([np.ones((len(batch_inds), 1)), not_dones[:, 1:]]) * (indices != self.pos)
```

Comment:
oh... you also included the not dones in there

"not_dones that are from other transitions are filtered out." -> not super clear to me....

and still not sure what the filter should look like...

Reply:
The idea is that the filter selects the relevant rewards. The filter always has col 0 to be 1, because it is the immediate reward and arange at col 0 is 0, so that is okay.

The not dones does the same, except for steps 1, 2, ... n-1, so it chooses which rewards to take into consideration. When done =1, `multiply.accumulate` zeros out the future rewards.

> "not_dones that are from other transitions are filtered out." -> not super clear to me....

What I meant to say was that `self.pos` points to either older or empty transitions. We need to account for this through the filter. So using `indices != pos` ensures that we ignore the empty or older transition. 

Did I clear things a bit?